### PR TITLE
Automatically convert URLs in values into links in the dashboard

### LINKED ISSUE #4845: Dashboard does not display URLs stored as env variables as links
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

URLs are only displayed as plain texts under the `Details` section of resources instead of being rendered as clickable links

### Expected Behavior

URLs should be displayed as clickable links instead of plain text

### Steps To Reproduce

1. Enable Aspire on ACA
2. Create an environment variable and store URL as it's value.
3. Open Aspire dashboard and should find the env variable as plain text under `Details`

### Exceptions (if any)

_No response_

### .NET Version info

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by JamesNK**
## Description

Automatically convert URLs in values and properties into links:

image

GridValue is used in many places:
* Resource properties, structured logs properties, span propertries
* Env vars
* Message on structured logs grid

There is a limitation: it doesn't work together with highlighting. Highlighting is only supported over plain text, not HTML. That means links are disabled when you begin searching with a filter and you go back to the current experience.

Getting these two features to work together properly is quite tricky. Will require building up a model of the text and transforming it when applying different operations (adding in links, adding in highlights)

Addresses https://github.com/dotnet/aspire/issues/6175
Fixes https://github.com/dotnet/aspire/issues/4845

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [x] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

### CODE REVIEW

**Comment by JamesNK** [comment]
This is an impactful change with some risk. Post 9.0 feature.

**Review by adamint — APPROVED** [review]

**Comment by JamesNK** [comment]
Please take another look. I came back to his and did some more testing and noticed that content wasn't properly HTML encoded. Required a change in order of operations when parsing logs and content.

File: src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs [hunk 1]
```diff
@@ -90,23 +96,44 @@ protected override void OnInitialized()
...
-        if (text is not null && MaxDisplayLength is int maxLength && text.Length > maxLength)
+        if (firstRender)
         {
-            return text[..maxLength];
+            // If the value and formatted value are different then there are hrefs in the text.
+            // Add a click event to the cell text that stops propagation if a href is clicked.
+            // This prevents details view from opening when the value is in a main page grid.
+            if (_value != _formattedValue)
+            {
+                await JS.InvokeVoidAsync("setCellTextClickHandler", _cellTextId);
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs:125]: I don't believe that works. The problem is the click event we want to stop propagating to is the grid row. That means we need to call `stopPropagation` before the event reaches the grid row. A common ancestor stopping propagation is too late.

Possible optimization: `stopPropagation` only needs to run on URLs inside data grids with row clicks. A flag could be added to only attach this event when needed.

[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs:125]: Makes sense. Maybe a `MutationObserver` could be used to still do all this in JS: https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver

**Comment by drewnoakes** [comment]
Please hold off on merging this for a day as I'm looking at https://github.com/dotnet/aspire/pull/6230 today for 9.0 and will want to backport it. There's overlap here which would make that backport difficult/risky.

File: src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs:134]: Added flag to opt-in to stop propagation.

File: src/Aspire.Dashboard/Components/Controls/GridValue.razor [hunk 2]
```diff
@@ -8,21 +8,21 @@
...
+        <span class="grid-value" title="@(ToolTip ?? Value)" id="@_cellTextId">
             @ContentBeforeValue
             @if (EnableHighlighting && !string.IsNullOrEmpty(HighlightText))
             {
                 <FluentHighlighter HighlightedText="@HighlightText" Text="@Value" />
             }
             else
             {
-                @Value
+                @((MarkupString)(_formattedValue ?? string.Empty))
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Components/Controls/GridValue.razor:25]: Could move the null check to the `else` above and simplify this a bit.

File: src/Aspire.Dashboard/ConsoleLogs/UrlParser.cs [hunk 3]
```diff
@@ -50,17 +52,29 @@ public static bool TryParse(string? text, [NotNullWhen(true)] out string? modifi
...
-    // /b                             - Match must start at a word boundary (after whitespace or at the start of the text)
+    // (?<=m|\\b)                     - Match must start at a word boundary (after whitespace or at the start of the text) or "m".
+    //                                  "m" is a trailing character of ANSI escape sequences.
+    //                                  Required because URLs are matched before processing ANSI escape sequences.
     // https?://                      - http:// or https://
     // [-A-Za-z0-9+&@#/%?=~_|$!:,.;]* - Any character in the list, matched zero or more times.
     // [A-Za-z0-9+&@#/%=~_|$]         - Any character in the list, matched exactly once
-    [GeneratedRegex("\\bhttps?://[-A-Za-z0-9+&@#/%?=~_|$!:,.;]*[A-Za-z0-9+&@#/%=~_|$]")]
-    private static partial Regex GenerateUrlRegEx();
+    [GeneratedRegex("(?<=m|\\b)https?://[-A-Za-z0-9+&@#/%?=~_|$!:,.;]*[A-Za-z0-9+&@#/%=~_|$]")]
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/ConsoleLogs/UrlParser.cs:78]: Looks like we don't use captured groups here, so we could add `RegexOptions.ExplicitCapture` to allocate less when matching, now that we've added a group.

**Review by drewnoakes — APPROVED** [review]

File: src/Aspire.Dashboard/ConsoleLogs/UrlParser.cs (hunk 3 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/ConsoleLogs/UrlParser.cs]: I looked at this regex more and I removed the group. I added flags to ignore case and culture.

**Pull request merged**